### PR TITLE
feat: refactor filters into standalone package

### PR DESCRIPTION
**Pull request by achrinza**
see: #5957

Signed-off-by: Rifa Achrinza <25147899+achrinza@users.noreply.github.com>

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by achrinza** [comment]
The coverage drop doesn't seem to be related to this PR.

File: packages/filter/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * Objects with open properties
+ */
+export interface AnyObject {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [property: string]: any;
+}
```
Comment (raymondfeng) [inline packages/filter/src/types.ts:12]: This can be replaced with `Record<string, any>`.

File: packages/repository/package.json [hunk 2]
```diff
@@ -33,6 +33,7 @@
   },
   "dependencies": {
     "@loopback/core": "^2.9.4",
+    "@loopback/filter": "~0.1.0",
```
Comment (raymondfeng) [inline packages/repository/package.json:36]: Let's use `^0.1.0`.

File: packages/filter/package.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "@loopback/filter",
+  "version": "0.1.0",
```
Comment (raymondfeng) [inline packages/filter/package.json:3]: Maybe we should start with `1.0.0`?

File: packages/filter/package.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  ],
+  "dependencies": {
+    "@loopback/core": "^2.9.4",
+    "tslib": "^2.0.0"
+  },
+  "devDependencies": {
+    "@loopback/build": "^6.2.1",
+    "@loopback/testlab": "^3.2.3",
+    "@types/node": "^10.17.28",
+    "typescript": "~3.9.7"
```
Comment (raymondfeng) [inline packages/filter/package.json:42]: `"typescript": "~4.0.2"`

**Comment by achrinza** [comment]
Thanks for the feedback, I agree with the points and applied them; PTAL.

File: packages/filter/README.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
+# @loopback/filter
```
Comment (raymondfeng) [inline packages/filter/README.md:1]: Which name is better? 

- `@loopback/filter`
- `@loopback/query-filter`
- `@loopback/repository-filter`

**Comment by raymondfeng** [comment]
@achrinza Good initiative. The change LGTM. Let's settle on the name - https://github.com/strongloop/loopback-next/pull/6182/files#r477425945.

File: packages/filter/README.md (hunk 5 above)
Comment (achrinza) [reply packages/filter/README.md:1]: I prefer `@loopback/filter` as the filter is use both when working with queries and repositories. Prefixing it with `query-` or `repository-` may create confusion that the filter only applies when working with either one of those concepts.

Comment (bajtos) [reply packages/filter/README.md:1]: Good discussion!

I vote for `@loopback/filter`.

File: packages/filter/package.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  "author": "IBM Corp.",
+  "license": "MIT",
+  "files": [
+    "README.md",
+    "dist",
+    "src",
+    "!*/__tests__"
+  ],
+  "dependencies": {
+    "@loopback/core": "^2.9.4",
```
Comment (bajtos) [inline packages/filter/package.json:35]: AFAICT, this new package does not need anything from `@loopback/core`. Can we remove this dependency please?

If not then please move it to `peerDependencies`, see #5959.

**Review by bajtos — CHANGES_REQUESTED** [review]
Great initiative, @achrinza!

The code looks mostly good, I have one comment to address 👇🏻

**Review by hacksparrow — APPROVED** [review]
👍

**Review by bajtos — APPROVED** [review]
👏🏻

**Pull request merged**